python/ray/autoscaler/v2/instance_manager/instance_manager.py

```diff
@@ -63,3 +66,205 @@ def get_instance_manager_state(
+            Instance.RAY_INSTALL_FAILED,
+        }
+
+    @staticmethod
+    def is_ray_running_reachable(instance: Instance) -> bool:
+        """
+        Returns True if the instance is in a status where it may transition
+        to RAY_RUNNING status.
+        """
+        assert instance.status != Instance.UNKNOWN
+        return instance.status in [
+            Instance.UNKNOWN,
+            Instance.QUEUED,
+            Instance.REQUESTED,
+            Instance.ALLOCATED,
+            Instance.RAY_INSTALLING,
+            Instance.ALLOCATION_FAILED,
+        ]
+
+    @staticmethod
+    def set_status(
+        instance: Instance,
+        new_instance_status: Instance.InstanceStatus,
+        details: str = "",
+    ):
+        """Transitions the instance to the new state.
+
+        Args:
+            instance: The instance to update.
+            new_instance_status: The new status to transition to.
+            details: The details of the transition.
+
+        Raises:
+            ValueError if the transition is not allowed.
+        """
+        if (
+            new_instance_status
+            not in InstanceUtil.get_valid_transitions()[instance.status]
+        ):
+            raise ValueError(
+                f"Invalid transition from {instance.status} to {new_instance_status}"
+            )
+        instance.status = new_instance_status
+        InstanceUtil._record_status_transition(instance, new_instance_status, details)
+
+    @staticmethod
+    def _record_status_transition(
+        instance: Instance, status: Instance.InstanceStatus, details: str
+    ):
+        """Records the status transition.
+
+        Args:
+            instance: The instance to update.
+            status: The new status to transition to.
+        """
+        now_ms = time.time_ns() // 1000000
+        instance.status_history.append(
+            Instance.StatusHistory(
+                instance_status=status,
+                timestamp_ms=now_ms,
+                details=details,
+            )
+        )
+
+    @staticmethod
+    def get_valid_transitions() -> Dict[
+        "Instance.InstanceStatus", Set["Instance.InstanceStatus"]
+    ]:
+        return {
+            Instance.QUEUED: {
+                # Cloud provider requested to launch a node for the instance.
+                Instance.REQUESTED
+            },
+            Instance.REQUESTED: {
+                # Cloud provider allocated a cloud node for the instance.
+                Instance.ALLOCATED,
+                # Cloud provider failed to allocate one. Either timeout or
+                # failed immediately.
+                Instance.ALLOCATION_FAILED,
+            },
+            Instance.ALLOCATED: {
+                # Ray needs to be install and launch on the provisioned cloud node.
+                Instance.RAY_INSTALLING,
+                # Ray is already installed and running on the provisioned cloud node.
+                Instance.RAY_RUNNING,
```

Comment:
Could you add comments about when each transition can happen? E.g. when will we transition from allocated directly to ray_running.